R/smoothDegreeDays.R

```diff
@@ -0,0 +1,131 @@
+
+  # PARAMETERS -----------------------------------------------------------------
+
+  # upper temporal threshold for historical data points
+  endOfHistory <- 2020
+
+
+
+  # PROCESS DATA ---------------------------------------------------------------
+
+  dataSmooth <- data %>%
+
+    # smooth model data before averaging to avoid impact of outliers
+    group_by(across(-all_of(c("period", "value")))) %>%
+    mutate(value = ifelse(.data[["period"]] <= endOfHistory |
+                            .data[["rcp"]] == "picontrol",
+                          .data[["value"]],
+                          lowpass(.data[["value"]], i = nSmoothIter))) %>%
+    ungroup() %>%
+
+    # take mean over models
+    group_by(across(-all_of(c("model", "value")))) %>%
+    reframe(value = mean(.data[["value"]])) %>%
+    ungroup()
+
+
+  # fit linear regression to last n (= nHistYears) values per model and create
+  # predictions for m (= transitionYears) periods
+  transitionPreds <- data %>%
+    filter(.data[["period"]] >= (endOfHistory - nHistYears),
+           .data[["period"]] <= endOfHistory,
+           .data[["rcp"]] == "historical",
+           !is.na(.data[["value"]])) %>%
+
+    group_by(across(all_of(c("region", "variable", "tlim", "model")))) %>%
+    group_modify(~{
+      # create prediction data frame
+      predPeriods <- data.frame(
+        period = seq(endOfHistory, endOfHistory + transitionYears, 1)
+      )
+
+      # linear fit
+      fit <- lm(value ~ period, data = .x)
+
+      # predict future values
+      predPeriods$prediction <- predict(fit, newdata = predPeriods)
+      return(predPeriods)
+    }) %>%
+    ungroup() %>%
+
+    # average predictions across models
+    group_by(across(-all_of(c("model", "prediction")))) %>%
+    reframe(prediction = mean(.data[["prediction"]])) %>%
+    ungroup()
+
+
+  # filter data w.r.t. periods
+  dataSmooth <- rbind(dataSmooth %>%
+                        filter(.data[["rcp"]] == "historical",
+                               .data[["period"]] <= endOfHistory),
+                      dataSmooth %>%
+                        filter(.data[["rcp"]] != "historical",
+                               .data[["period"]] >= endOfHistory))
+
+
+  # smooth transition between last historical value and projections to minimize
+  # deviations caused by the preceding smoothing
+  dataSmooth <- dataSmooth %>%
+    left_join(transitionPreds,
+              by = c("region", "variable", "period")) %>%
+    mutate(
+      value = case_when(
+        .data[["period"]] >= endOfHistory &
+          .data[["period"]] <= (endOfHistory + transitionYears) &
+          rcp != "picontrol" & !is.na(.data[["prediction"]]) & !is.na(.data[["value"]]) ~
+          .data[["prediction"]] + (.data[["value"]] - .data[["prediction"]]) *
+            ((.data[["period"]] - endOfHistory) / transitionYears),
+        TRUE ~ .data[["value"]]
+      )
```

Comment:
If you only have one consition, I'd finde `ifelse` easier to read.
```suggestion
      value = ifelse(
        .data[["period"]] >= endOfHistory &
          .data[["period"]] <= (endOfHistory + transitionYears) &
          rcp != "picontrol" & !is.na(.data[["prediction"]]) & !is.na(.data[["value"]]),
        .data[["prediction"]] + (.data[["value"]] - .data[["prediction"]]) *
          ((.data[["period"]] - endOfHistory) / transitionYears),
        .data[["value"]]
      )
```